internal/chooserbenchmark/client.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package chooserbenchmark
+
+import (
+	"context"
+	"math/rand"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+	"go.uber.org/net/metrics/bucket"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/peer"
+)
+
+var (
+	BucketMs  = bucket.NewRPCLatency()
+	BucketLen = len(BucketMs)
+)
+
+type Client struct {
+	groupName  string
+	id         int
+	reqCounter int
+	resCounter atomic.Int32
+	histogram  []atomic.Int32
+	mu         float64
+	sigma      float64
+	chooser    *peer.BoundChooser
+	start      chan struct{}
+	stop       chan struct{}
+	wg         *sync.WaitGroup
+	listeners  *Listeners
+}
+
+func PeerListChooser(constructor PeerListConstructor, serverCount int) *peer.BoundChooser {
+	return peer.Bind(constructor(NewBenchTransport()), peer.BindPeers(NewPeerIdentifiers(serverCount)))
+}
+
+func NewClient(id int, group *ClientGroup, listeners *Listeners, start, stop chan struct{}, wg *sync.WaitGroup, constructor PeerListConstructor, serverCount int) *Client {
+	plc := PeerListChooser(constructor, serverCount)
+	sleepTime := float64(time.Second) / float64(group.RPS)
+	return &Client{
+		groupName: group.Name,
+		id:        id,
+		histogram: make([]atomic.Int32, BucketLen),
+		mu:        sleepTime,
+		sigma:     sleepTime / 20,
+		chooser:   plc,
+		start:     start,
+		stop:      stop,
+		wg:        wg,
+		listeners: listeners,
+	}
+}
+
+func (c *Client) normalSleepTime() time.Duration {
+	return time.Duration(c.sigma*rand.NormFloat64() + c.mu)
+}
+
+func (c *Client) incBucket(t time.Duration) {
```

Comment:
Yes, have a `clientHistogram` would be good. But I just use `histogram`, the number fields will not be reduced. I'm not sure whether to create a new type, because many things like `metrics`, `signals`, etc, could be decoupled. That would make things complicated. Waiting for your follow-ups.